### PR TITLE
Allow CBMC to call an external sat solver

### PR DESCRIPTION
**Pull request by fbbotero-aws**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/cbmc/cbmc_parse_options.cpp [hunk 1]
```diff
@@ -1109,6 +1116,7 @@ void cbmc_parse_optionst::help()
     " --yices                      use Yices\n"
     " --z3                         use Z3\n"
     " --refine                     use refinement procedure (experimental)\n"
+    " --sat-solver-invocation cmd  command to invoke SAT solver process (experimental)\n"
```
Comment (martin-cs) [inline src/cbmc/cbmc_parse_options.cpp]: Does this need to be experimental?

File: src/solvers/sat/external_sat.cpp [hunk 2]
```diff
@@ -0,0 +1,188 @@
+/*******************************************************************\
+
+Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (hannes-steffenhagen-diffblue) [inline src/solvers/sat/external_sat.cpp]: :no_entry_sign: This is currently saying that the copyright is either owned by Amazon or someone who isn’t Amazon, which is pretty confusing. The all rights reserved bit is also a bit weird, if you are making a contribution it must be under the same license as cbmc itself (see https://github.com/diffblue/cbmc/blob/develop/LICENSE).

Also instead of heaving a header like this please specify who the author or copyright owner is via doxygen using syntax like

```
/// \file
/// \author author-names-go-here
```

File: src/goto-checker/solver_factory.cpp [hunk 3]
```diff
@@ -229,9 +232,25 @@ std::unique_ptr<solver_factoryt::solvert> solver_factoryt::get_dimacs()
...
   return util_make_unique<solvert>(std::move(bv_dimacs), std::move(prop));
 }
 
+std::unique_ptr<solver_factoryt::solvert> solver_factoryt::get_external_sat()
+{
+  no_beautification();
+  no_incremental_check();
+
+  std::string sat_solver_invocation =
+    options.get_option("sat-solver-invocation");
```
Comment (martin-cs) [inline src/goto-checker/solver_factory.cpp]: I feel it would be nicer if this was an argument, thus pushing the use of `options` 'higher up' in the call stack.  `get_smt2` takes an argument so this would not be entirely inconsistent, also the place where this is called is likely to already reference `options`.

File: src/solvers/sat/external_sat.cpp [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+{
+  UNIMPLEMENTED;
+}
+
+void external_satt::set_assignment(literalt, bool)
+{
+  UNIMPLEMENTED;
+}
+
+inline void external_satt::write_cnf_file(std::string cnf_file)
```
Comment (hannes-steffenhagen-diffblue) [inline src/solvers/sat/external_sat.cpp]: :no_entry_sign: `inline` is redundant on class methods which are defined in-place

File: src/solvers/sat/dimacs_cnf.h [hunk 5]
```diff
@@ -32,6 +32,9 @@ class dimacs_cnft:public cnf_clause_listt
   void set_assignment(literalt a, bool value) override;
   bool is_in_conflict(literalt l) const override;
 
+  static void
```
Comment (martin-cs) [inline src/solvers/sat/dimacs_cnf.h:35]: Why `static`? In fact, why this change at all?

File: src/solvers/sat/external_sat.cpp [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+#include <chrono>
+#include <cstdlib>
+#include <fstream>
+#include <random>
+#include <sstream>
+#include <string>
+#include <thread>
+
+class missing_configurationt : public cprover_exception_baset
```
Comment (martin-cs) [inline src/solvers/sat/external_sat.cpp]: This feels like something that could be a bit more general.

File: src/solvers/sat/external_sat.cpp [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+const std::string external_satt::solver_text()
+{
+  return "External SAT solver";
+}
+
+bool external_satt::is_in_conflict(literalt) const
+{
+  UNIMPLEMENTED;
```
Comment (martin-cs) [inline src/solvers/sat/external_sat.cpp:35]: The original intent of `UNIMPLEMENTED` was to document incomplete parts of the code.  Ideally we weren't going to add more.  That said, it is not clear to me what would be better here.

**Review by martin-cs — COMMENTED** [review]
Given that most, if not all of MiniSAT is built during the normal build procedure, would it be possible to have some tests for this?

File: src/solvers/sat/external_sat.cpp (hunk 6 above)
Comment (hannes-steffenhagen-diffblue) [inline src/solvers/sat/external_sat.cpp]: This isn’t thrown anywhere right now. If I’m guessing the intent for where this is eventually supposed to be correctly this should probably be something like `invalid_command_line_argument_exceptiont`

File: src/solvers/sat/external_sat.cpp [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+  {
+    return _message;
+  }
+
+protected:
+  std::string _message;
+};
+
+external_satt::external_satt(message_handlert &message_handler, std::string cmd)
+  : cnf_clause_list_assignmentt(message_handler), _cmd(std::move(cmd))
```
Comment (hannes-steffenhagen-diffblue) [inline src/solvers/sat/external_sat.cpp]: Don’t start members with an underscore.

**Review by hannes-steffenhagen-diffblue — COMMENTED** [review]
This needs some more documentation, like what exact commandline interface is this expecting for external solvers? In what format is the formula presented, what does this code expect to be given back?

File: src/cbmc/cbmc_parse_options.h [hunk 9]
```diff
@@ -57,6 +57,7 @@ class optionst;
   OPT_JSON_INTERFACE \
   "(smt1)(smt2)(fpa)(cvc3)(cvc4)(boolector)(yices)(z3)(mathsat)" \
   "(cprover-smt2)" \
+  "(sat-solver-invocation):" \
```
Comment (peterschrammel) [inline src/cbmc/cbmc_parse_options.h]: Is this a good name? What about `--external-sat-solver` ?

File: src/goto-checker/solver_factory.cpp (hunk 3 above)
Comment (peterschrammel) [reply src/goto-checker/solver_factory.cpp]: `solver_factoryt` relies heavily on `options` so this is fine IMO.

File: src/solvers/sat/external_sat.h [hunk 10]
```diff
@@ -0,0 +1,42 @@
+/*******************************************************************\
+
+Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (peterschrammel) [inline src/solvers/sat/external_sat.h]: 🐰

File: src/solvers/sat/external_sat.cpp (hunk 2 above)
Comment (peterschrammel) [reply src/solvers/sat/external_sat.cpp]: 🐰

**Review by peterschrammel — COMMENTED** [review]
It would be good to have tests for this.

File: src/goto-checker/solver_factory.cpp [hunk 11]
```diff
@@ -229,9 +232,25 @@ std::unique_ptr<solver_factoryt::solvert> solver_factoryt::get_dimacs()
 
   auto bv_dimacs =
     util_make_unique<bv_dimacst>(ns, *prop, message_handler, filename);
+
   return util_make_unique<solvert>(std::move(bv_dimacs), std::move(prop));
 }
 
+std::unique_ptr<solver_factoryt::solvert> solver_factoryt::get_external_sat()
+{
+  no_beautification();
```
Comment (peterschrammel) [inline src/goto-checker/solver_factory.cpp:241]: 👍

File: src/solvers/sat/dimacs_cnf.h (hunk 5 above)
Comment (fbbotero-aws) [reply src/solvers/sat/dimacs_cnf.h:35]: It gets reused when we write out the CNF file in external_sat.cpp
https://github.com/diffblue/cbmc/pull/5469/files#diff-34efc04d56d4073dfc15a3cf4e548772R52

File: src/solvers/sat/external_sat.cpp (hunk 6 above)
Comment (fbbotero-aws) [reply src/solvers/sat/external_sat.cpp]: removed it

File: src/solvers/sat/external_sat.cpp (hunk 8 above)
Comment (fbbotero-aws) [reply src/solvers/sat/external_sat.cpp]: Changed it to solver_cmd

**Comment by fbbotero-aws** [comment]
> This needs some more documentation, like what exact commandline interface is this expecting for external solvers? In what format is the formula presented, what does this code expect to be given back?

Where would you like this documentation? In the code or online help?

**Comment by fbbotero-aws** [comment]
> Given that most, if not all of MiniSAT is built during the normal build procedure, would it be possible to have some tests for this?

MiniSAT doesn't produce output compatible with flag. We used the output generated as part of SAT comp, which seems to have become pretty standard across multiple solvers. The format is generally:

```
s <SATISFIABILITY RESULT: SATISFIABLE, UNSATISFIABLE>
v <SATISFYING ASSIGNMENTS>
v <SATISFYING ASSIGNMENTS CAN BE ON MULTIPLE LINES>
```

Minisat currently produces:
```
minisat satis.cnf -verb=0 /dev/stdout
WARNING: for repeatability, setting FPU to use double precision
SATISFIABLE
SAT
1 -2 0
```

I can write some unit tests around the parsing

**Review by peterschrammel — APPROVED** [review]

**Comment by peterschrammel** [comment]
The coverage drop is not spurious. There simply no test for the code that you have added. Can you add one?

**Comment by NlightNFotis** [comment]
@fbbotero-aws Is it possible to address the above comment?

If it's resolved, I will merge it after that.

**Comment by fbbotero-aws** [comment]
> @fbbotero-aws Is it possible to address the above comment?
> 
> If it's resolved, I will merge it after that.

I'm adding tests to the results parser to increase coverage

**Comment by fbbotero-aws** [comment]
@NlightNFotis @peterschrammel Added some coverage.

**Comment by fbbotero-aws** [comment]
Looks like one of the checks timed out. Is there a way to rerun a travis job?

**Comment by martin-cs** [comment]
Some people have the power to restart Travis, failing that a push (with rebase or insignificant change) should re-trigger it.

**Comment by martin-cs** [comment]
@fbbotero-aws There have been changes to develop so if you just rebase and force push then everything should work.

**Comment by NlightNFotis** [comment]
I don't see this having conflicts - it's on my todo for later today/early tomorrow to inspect the tests and merge once having done so. @martin-cs @fbbotero-aws
